libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,247 @@
+#include <eosio/reliable_amqp_publisher/reliable_amqp_publisher.hpp>
+
+#include <fc/network/url.hpp>
+#include <fc/exception/exception.hpp>
+#include <fc/io/cfile.hpp>
+#include <fc/log/logger_config.hpp> //set_os_thread_name()
+
+#include <amqpcpp.h>
+#include <amqpcpp/libboostasio.h>
+
+#include <thread>
+#include <chrono>
+#include <future>
+#include <algorithm>
+
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/filesystem.hpp>
+
+namespace eosio {
+
+struct reliable_amqp_publisher_callbacks {
+   virtual void amqp_ready() = 0;
+   virtual void amqp_error(const char* msg) = 0;
+};
+
+struct reliable_amqp_publisher_handler : AMQP::LibBoostAsioHandler {
+   reliable_amqp_publisher_handler(reliable_amqp_publisher_callbacks* callbacks, boost::asio::io_context& ctx) :
+      AMQP::LibBoostAsioHandler(ctx),
+      cbs(callbacks) {}
+
+   void onReady(AMQP::TcpConnection* c) override {
+      cbs->amqp_ready();
+   }
+
+   //disable heartbeat for now
+   uint16_t onNegotiate(AMQP::TcpConnection* connection, uint16_t interval) override {
+      return 0;
+   }
+
+   void onError(AMQP::TcpConnection*, const char* message) override {
+      cbs->amqp_error(message);
+   }
+
+   reliable_amqp_publisher_callbacks* cbs;
+};
+
+struct reliable_amqp_publisher_impl final : reliable_amqp_publisher_callbacks {
+   reliable_amqp_publisher_impl(const std::string url, const std::string exchange, const std::string routing_key, const boost::filesystem::path unconfirmed_path, const std::optional<std::string> message_id) :
```

Comment:
I was probably thinking passing by reference implies the passed object should remain valid for the entire life of the created object.

Reply:
I didn't think that const reference arguments normally had such an implication.  In any case, if you do pass by value, the arguments should be moved instead of being copied.